src/vset.c

```diff
@@ -0,0 +1,2094 @@
+ * `capacity` elements. Internally, it delegates allocation and setup to
+ * `pvMakeRoomFor`, starting from a NULL vector.
+ *
+ * Arguments:
+ *   capacity - The initial number of elements the vector should be able to store.
+ *
+ * Return:
+ *   A pointer to the newly allocated pVector.
+ *   Note that a NULL is a !!valid!! cector which size is zero.
+ *
+ * Note:
+ *   The logical length (`len`) of the returned vector is initialized to 0.
+ */
+pVector *pvNew(uint32_t capacity) {
+    pVector *new_vec = NULL;
+    return pvMakeRoomFor(new_vec, capacity);
+}
+
+/* Inserts an element at the specified position in the pVector.
+ *
+ * Ensures enough capacity for the new element, shifts elements to make space,
+ * and inserts the given element at the desired position.
+ *
+ * Arguments:
+ *   pv   - The pVector to insert into (can be NULL).
+ *   elem - The pointer to be inserted.
+ *   pos  - The index at which to insert the element (must be ≤ pv->len).
+ *
+ * Return:
+ *   The updated pVector with the element inserted. */
+pVector *pvInsert(pVector *pv, void *elem, uint32_t pos) {
+    pv = pvMakeRoomFor(pv, 1);
+
+    if (pos < pv->len) {
+        memmove(&pv->data[pos + 1], &pv->data[pos], (pv->len - pos) * sizeof(void *));
+    }
+
+    pv->data[pos] = elem;
+    pv->len++;
+    return pv;
+}
+
+/* Removes the element at the specified index from the pVector.
+ *
+ * Shifts elements as necessary and optionally shrinks the vector if memory can be saved.
+ * If this is the last element in the vector, the vector is freed and NULL is returned.
+ *
+ * Arguments:
+ *   pv  - The pVector to remove from.
+ *   idx - The index of the element to remove (must be < pv->len).
+ *
+ * Return:
+ *   The updated pVector after removal.
+ *   Returns NULL if the last element was removed and the vector was freed. */
+pVector *pvRemoveAt(pVector *pv, uint32_t idx) {
+    assert(pv && pv->len > 0);
+    assert(idx < pv->len);
+    if (pv->len == 1) {
+        /* Last element being removed; delete vector */
+        zfree(pv);
+        return NULL;
+    } else if (idx < pv->len - 1UL)
+        memmove(&pv->data[idx], &pv->data[idx + 1], (pv->len - idx - 1) * PV_ELEM_SIZE);
+    pv->len--;
+    return pvShrinkToFit(pv);
+}
+
+/* Removes the first matching element from the pVector.
+ *
+ * Performs a linear search for the given pointer and removes the first match.
+ * Updates the vector pointer in case a removal was done.
+ *
+ * Arguments:
+ *   pv   - A pointer to the location of the pVector to remove from.
+ *   elem - The element pointer to match and remove.
+ *
+ * Return:
+ *   true in case a removal was made, false otherwise */
+bool pvRemove(pVector **ppv, void *elem) {
+    pVector *pv = *ppv;
+    if (!pv || pv->len == 0) return false;
+
+    for (uint32_t i = 0; i < pv->len; i++) {
+        if (pv->data[i] == elem) {
+            *ppv = pvRemoveAt(pv, i);
+            return true;
+        }
+    }
```

Comment:
DRY.  `pvFind` duplicates most of this code.  Suggest moving `pvFind` before this function (to avoid the forward declare).  And then have this function call `pvFind` then `pvRemoveAt`.